### PR TITLE
get rid of snapshotLength and hasNewContent

### LINKED ISSUE #1567: refactor: get rid of snapshotLength and hasNewContent
Spin off of #1494 

SnapshotLength and hasNewContent  reflect a design problem: with those methods, DJPP depends on the printing state to reason about what to do. We have a dedicated concept for this, the PrintingContext. It now seems to me that we should remove snapshotLength and hasNewContent (whether char-based or token-based) and add some logic in PrintingContext instead.

### PR DESCRIPTION
**Pull request by monperrus**
fix #1567

### CODE REVIEW

**Comment by pvojtechovsky** [comment]
I wonder that conditions detecting whether `.` is needed are so simple! And all tests passed! You know the spoon really well!!! ;-)
Thank you @monperrus

**Comment by monperrus** [comment]
I was myself quite surprised that it works out-of-the-box!

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 1]
```diff
@@ -815,9 +815,9 @@ public void visitCtDo(CtDo doLoop) {
 							target.setImplicit(false);
 						}
 					}
-					printer.getPrinterHelper().snapshotLength();
 					scan(target);
-					if (printer.getPrinterHelper().hasNewContent()) {
+					// the implicit drives the separator
+					if (!target.isImplicit()) {
 						printer.writeSeparator(".");
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: move scan call above into this if block?
Then it is clear when it produces output.

**Comment by pvojtechovsky** [comment]
And please uncomment the list line of token writer printer test. It should pass now.

File: src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java [hunk 2]
```diff
@@ -1099,25 +1098,23 @@ public void visitCtIf(CtIf ifElement) {
 			if (parentType != null && parentType.getQualifiedName() != null && parentType.getQualifiedName().equals(invocation.getExecutable().getDeclaringType().getQualifiedName())) {
 				printer.writeKeyword("this");
 			} else {
-				printer.getPrinterHelper().snapshotLength();
 				scan(invocation.getTarget());
-				if (printer.getPrinterHelper().hasNewContent()) {
+				if (invocation.getTarget() != null && !invocation.getTarget().isImplicit()) {
 					printer.writeSeparator(".");
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java]: What about this call of scan?

**Comment by pvojtechovsky** [comment]
It is OK for me. Let me know if you are finished. I will merge it then.

**Comment by monperrus** [comment]
I'm done, you can merge, thanks.

**Pull request merged**